src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,407 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relax/transform/gradient.cc
+ * \brief A reverse-mode automatic differentiation. For a given function specified by the input
+ * global var, it generates a new function which computes the adjoints of the given function's
+ * inputs in respect of the function's output. Use require_grads to specify inputs which need to be
+ * differentiated instead of returning adjoints of all inputs.
+ *
+ * Now only supports differentiating a function in the IRModule with one dataflow block
+ * with respect to the only return value of the function. It needs to be scalar.
+ *
+ * For examples, see the MLP examples in tests/python/relax/test_transform_gradient.py and
+ * tests/python/relax/test_transform_gradient_numeric.py.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include <unordered_set>
+
+namespace tvm {
+namespace relax {
+
+class GradientMutator : public ExprMutator {
+ public:
+  explicit GradientMutator(IRModule mod, const Array<Var>& require_grads)
+      : ExprMutator(std::move(mod)), require_grads_(std::move(require_grads)) {}
+
+  Function FuncTransform(Function func) {
+    ICHECK(func->body->IsInstance<SeqExprNode>());
+
+    Array<Var> new_params;
+    for (Var param : func->params) {
+      Var new_param = Var(param->vid, param->shape(), param->checked_type(), param->span);
+      this->var_remap_[param->vid] = new_param;
+      new_params.push_back(new_param);
+    }
+    Expr new_body = VisitWithNewScope(func->body);
+
+    const SeqExprNode* seq_expr = new_body.as<SeqExprNode>();
+    // only support a single dataflow block
+    ICHECK(seq_expr->blocks.size() == 1);
+    ICHECK(seq_expr->blocks[0]->IsInstance<DataflowBlockNode>());
+    const DataflowBlockNode* block = seq_expr->blocks[0].as<DataflowBlockNode>();
+
+    builder_->BeginDataflowBlock();
+    for (const auto& binding : block->bindings) {
+      VisitBinding(binding);
+    }
+
+    // create adjoint var for inputs
+    for (size_t i = 0; i < new_params.size(); ++i) {
+      if (require_grads_.empty() || std::find(require_grads_.begin(), require_grads_.end(),
+                                              func->params[i]) != require_grads_.end()) {
+        CreateAdjointVar(new_params[i], false);
+      } else {
+        CreateAdjointVar(new_params[i], true);
+      }
+    }
+
+    // the return value of the function will be treated as target
+    if (const auto* node = seq_expr->body.as<VarNode>()) {
+      const Var& target = GetRef<Var>(node);
+      CheckTarget(target);
+      target_var_ = target;
+    } else {
+      LOG(FATAL) << "the body of the function is not a relax.Var";
+    }
```

Comment:
Remember that you always have such utilities to use.
```suggestion
    const auto* node = seq_expr->body.as<VarNode>();
    CHECK(node != nullptr) << "the body of the function is not a relax.Var";
    const Var& target = GetRef<Var>(node);
    CheckTarget(target);
    target_var_ = target;
```

Reply:
OK. There are many places in this code needed to be changed to this pattern.

Reply:
No hurry at all. I haven’t yet went into the mutation logic. You can have a good sleep first.

Reply:
For such cases https://github.com/mlc-ai/relax/blob/75081dbda4f0d8ac4e4c3a826bf60294e1b45989/src/relax/transform/gradient.cc#L155-L185
reporting error in the end is fine, since we have multiple branches here.

For single-branch cases, we can definitely use the family of CHECKs.

Reply:
Thank you! I will go rest for now and tomorrow Yixin or I will fix remaining comments. Merry Xmas

Reply:
Thanks and you too! I’m thinking about having you guiding me through the logic tomorrow some time, preferably your evening (since I don’t think you can get up before I sleep lol). Let’s coordinate tomorrow.